internal/config/config_file.go

```diff
@@ -8,23 +8,71 @@ import (
 	"os"
 	"path"
 	"path/filepath"
+	"runtime"
 	"syscall"
 
 	"github.com/mitchellh/go-homedir"
 	"gopkg.in/yaml.v3"
 )
 
 const (
-	GH_CONFIG_DIR = "GH_CONFIG_DIR"
+	GH_CONFIG_DIR   = "GH_CONFIG_DIR"
+	XDG_CONFIG_HOME = "XDG_CONFIG_HOME"
 )
 
+// Config path precedence
+// 1. GH_CONFIG_DIR
+// 2. XDG_CONFIG_HOME
+// 3. AppData (windows only)
+// 4. HOME
 func ConfigDir() string {
-	if v := os.Getenv(GH_CONFIG_DIR); v != "" {
-		return v
+	var path string
+	if a := os.Getenv(GH_CONFIG_DIR); a != "" {
+		path = a
+	} else if b := os.Getenv(XDG_CONFIG_HOME); b != "" {
+		path = filepath.Join(b, "gh")
+	} else if c, err := os.UserConfigDir(); runtime.GOOS == "windows" && err == nil && c != "" {
+		path = filepath.Join(c, "GitHub CLI")
+	} else {
+		d, _ := os.UserHomeDir()
+		path = filepath.Join(d, ".config", "gh")
+	}
+
+	// If the path does not exist try migrating config from default paths
+	if _, err := os.Stat(path); err != nil {
+		autoMigrateConfigDir(path)
 	}
 
-	homeDir, _ := homeDirAutoMigrate()
-	return homeDir
+	return path
+}
+
+// Check default paths (os.UserHomeDir, and homedir.Dir) for existing configs
+// If configs exist then move them to newPath
+// TODO: Remove support for homedir.Dir location in v2
+func autoMigrateConfigDir(newPath string) {
+	var paths []string
+	p1, _ := os.UserHomeDir()
+	paths = append(paths, p1)
+	p2, _ := homedir.Dir()
```

Comment:
We moved away from `homedir.Dir()` because it can be a pretty expensive operation (take a look at its implementation; it does a lot!). Even though we keep backwards compatibility with it, we used to avoid invoking it unless it was the absolutely last filesystem location that we tried. The old `homeDirAutoMigrate` function was specifically designed around this approach. This removes the guard and unconditionally calls `homedir.Dir()` even though the path returned in `p1` might be valid and might get used.

Couldn't we just keep using `homeDirAutoMigrate` intact and just migrate the result of that to the new location? I.e. to wrap the old migrate in a new migrate function.

Reply:
That is good context. I was wondering why the implementation of `homeDirAutoMigrate` was designed that way. I can restore that behavior.